### DISCUSSION TITLE
How can I make an animation have a consistent speed rather than duration?

### DISCUSSION THREAD

**Discussion by trevorwrightdev** [motiondivision/motion#1553] [Q&A]
I am making a JS game where the player moves to the location that you click on. But right now, the only way I can adjust the speed of the player movement is by setting the duration of the animation. This isn't good because I don't want every distance travelled to take up the same duration. I want it to be based on velocity and distance. For example, traveling 50 pixels in distance should take one fourth of the time to travel 200 pixels in distance. Here is my current code for the animation:
<img width="247" alt="image" src="https://user-images.githubusercontent.com/86077703/170790548-dba0f684-8ddd-4221-9496-637819f01396.png">
Is there a way I can change this to make it do what I want? I tried using inertia animations but the player would move in a strange direction when that was the case. I also want no bouncing or anything like that. Thank you!

**Comment by harshagarwalnyu** [comment]
Calculate the distance between current and target positions, then derive duration from distance divided by desired speed:

```jsx
import { motion, useAnimation } from 'framer-motion';

const SPEED = 300; // pixels per second

function Player({ position }) {
  const controls = useAnimation();
  const currentPos = useRef({ x: 0, y: 0 });

  async function moveTo(targetX, targetY) {
    const dx = targetX - currentPos.current.x;
    const dy = targetY - currentPos.current.y;
    const distance = Math.hypot(dx, dy);
    const duration = distance / SPEED;

    await controls.start({
      x: targetX,
      y: targetY,
      transition: {
        duration,
        ease: 'linear',  // constant speed requires linear ease
      },
    });

    currentPos.current = { x: targetX, y: targetY };
  }

  return (
    <motion.div
      animate={controls}
      onClick={(e) => moveTo(e.clientX, e.clientY)}
    />
  );
}
```

**Key points:**
- `ease: 'linear'` is required — any other easing curve (`easeInOut`, spring) makes speed non-constant even with correct duration
- Track `currentPos` with a ref (not state) to avoid stale closure issues
- `Math.hypot(dx, dy)` = Euclidean distance (Pythagoras theorem)
- If you need deceleration near the target (game feel): use `ease: [0.2, 0, 0.8, 1]` Bezier but accept that speed won't be perfectly constant
- For physics-based feel without spring: use `useVelocity` + `useTransform` to read current velocity and interpolate smoothly

**Alternative — spring physics:**
```js
transition: { type: 'spring', mass: 1, stiffness: 400, damping: 40 }
```
Springs don't have constant speed but feel natural for character movement. Adjust `stiffness`/`damping` for responsiveness.
